README.md
# glrestore (Glacier Restore)

Easily restore objects from AWS S3 glacier from the command line. Built using python

## Installation options

### pip
```
$ pip install glrestore
```

## Quick start

### Show program help and modules:
```
$ glrestore -h
```

### Example command to restore some files for 7 days as quickly as possible:
```
$ glrestore -f s3://cool-bucket/users/mattolm/archived-*.csv -d 7 -s Expedited
```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project (attempts to) adhere to [Semantic Versioning](http://semver.org/).

## [1.1.1] - 2022-08-27
- Check the "wait" every 5 min, not constantly
- Add another line in the cost estimation

## [1.1.0] - 2022-05-19
- Fix cost reporting bug
- Add ability to pass a list of files to restore
- Add the "--wait" argument

## [1.0.1] - 2022-04-22
- Fix cost calculation bug

## [1.0.0] - 2022-04-21
- Added the ability to make a report

## [0.4.1] - 2022-04-21
- Bugfix

## [0.4.0] - 2022-04-21
- Speed increase

## [0.3.2] - 2022-04-21
- Speed increase

## [0.3.1] - 2022-04-21
- Speed increase

## [0.3.0] - 2022-04-21
- Add cost estimations

## [0.2.3] - 2022-04-18
- Add DEEP_ARCHIVE support

## [0.2.2] - 2022-04-18
-Bugfix

## [0.2.1] - 2022-04-18
- Dont try to restore things that cant be restored

## [0.2.0] - 2022-03-30
- Working functionality
- Tests working
- PyPi properly set up

## [0.1.1] - 2022-03-30
- Skeleton made

## [0.1.0] - 2022-03-30
- Project is born

test/test.py
"""
Tests for glrestore
"""

import os
import sys
import time
import shutil
import pytest
import importlib
import logging
import subprocess
import pandas as pd
from time import sleep
from threading import Thread

import glrestore
import glrestore.s3_utils

"""
SET UP TESTING UTILITIES
"""
class TestingClass():
    def teardown(self):
        importlib.reload(logging)
        if os.path.isdir(self.test_dir):
            shutil.rmtree(self.test_dir)
        os.mkdir(self.test_dir)
        #os.rmdir(self.s3_testing_loc)
        importlib.reload(logging)

@pytest.fixture()
def BTO():
    """
    Basic test object
    This object makes no copies of anything; just has references and does setup / cleanup

    aws s3 cp --profile sonn ~/Programs/inStrain/test/test_data/N5_271_010G1_scaffold_min1000.fa s3://sonn-current/users/mattolm/testing_house/glrestore/N5_271_010G1_scaffold_min1000.fa --storage-class GLACIER

    aws s3 cp --profile sonn ~/Programs/inStrain/test/test_data/N5_271_010G1_scaffold_min1000.fa s3://sonn-current/users/mattolm/testing_house/glrestore/rest_N5_271_010G1_scaffold_min1000.fa --storage-class GLACIER
    aws s3api restore-object --profile sonn --bucket sonn-current --key users/mattolm/testing_house/glrestore/rest_N5_271_010G1_scaffold_min1000.fa --restore-request '{"Days":300,"GlacierJobParameters":{"Tier":"Expedited"}}'
    """
    # Set up
    self = TestingClass()

    # Specify some s3 locations
    self.s3_testing_loc = "s3://sonn-current/users/mattolm/testing_house/glrestore/temp/"

    # Specify some file locations
    self.glacerized_file_loc = "s3://sonn-current/users/mattolm/testing_house/glrestore/N5_271_010G1_scaffold_min1000.fa"
    self.non_glacerized_file_loc = "s3://sonn-highavail/testing/ng_N5_271_010G1_scaffold_min1000.fa"
    self.restored_file_loc = "s3://sonn-current/users/mattolm/testing_house/glrestore/rest_N5_271_010G1_scaffold_min1000.fa"

    self.glacerized_file_list = os.path.join(str(os.getcwd()), 'testfiles/glacerizedfile.txt')

    self.test_dir = os.path.join(str(os.getcwd()), 'temp_testdir/')

    self.teardown()
    yield self
    self.teardown()

def run_glrestore(cmd):
    """
    Simulate as if you're calling glrestore from the command line
    """
    import glrestore.glrestore
    from unittest.mock import patch
    with patch.object(sys, 'argv', cmd.split(" ")):
        glrestore.glrestore.main()

def upload_glacierized_file():
    cmd = "aws s3 cp ~/Programs/inStrain/test/test_data/N5_271_010G1_scaffold_min1000.fa s3://sonn-current/users/mattolm/testing_house/glrestore/N5_271_010G1_scaffold_min1000.fa --storage-class GLACIER"
    subprocess.call(cmd, shell=True)

"""
UNIT TESTS
"""

def test_object_glacerized_v2(BTO):
    """
    test the "s3_utils.glacier_status" function
    """
    upload_glacierized_file()

    # Test ability to check non-glacierized file
    assert glrestore.s3_utils.glacier_status_v2(BTO.non_glacerized_file_loc) == 'no-glacier'

    # Test ability to check glacierized file
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-no-restore'

    # Test ability to recognize *restored* glacierized file
    assert glrestore.s3_utils.glacier_status_v2(BTO.restored_file_loc) == 'glacier-restored'

    # Test ability to recognize *restoring* glacierized file
    cmd = "aws s3api restore-object --bucket sonn-current --key users/mattolm/testing_house/glrestore/N5_271_010G1_scaffold_min1000.fa --restore-request '{\"Days\":1,\"GlacierJobParameters\":{\"Tier\":\"Expedited\"}}'"
    subprocess.call(cmd, shell=True)
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-restoring'

def test_classify_glacier_objects(BTO):
    """
    test the "s3_utils.classify_glacier_objects" function on individual files
    """
    pd.set_option('display.max_colwidth', None)
    pd.set_option('display.max_columns', None)

    upload_glacierized_file()

    db = glrestore.s3_utils.get_object_storage_class_v2(
        [BTO.non_glacerized_file_loc, BTO.glacerized_file_loc, BTO.restored_file_loc])

    assert len(db) == 3
    assert db['storage_class'].value_counts()['GLACIER'] == 2

def test_classify_glacier_objects_2(BTO):
    """
    test the "s3_utils.classify_glacier_objects" function on folders
    """
    pd.set_option('display.max_colwidth', None)
    pd.set_option('display.max_columns', None)

    upload_glacierized_file()

    db = glrestore.s3_utils.get_object_storage_class_v2(
        's3://sonn-current/users/mattolm/testing_house/glrestore/')

    assert len(db) == 3
    assert db['storage_class'].value_counts()['GLACIER'] == 3


"""
INTEGRATED TESTS
"""

def test_glrestore_1(BTO):
    """
    A simple test restoring a single object to new location
    """
    upload_glacierized_file()

    # Make sure this file starts off glacierized
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-no-restore'

    cmd = f"glrestore -f {BTO.glacerized_file_loc} --debug"
    run_glrestore(cmd)

    # Make sure this file is now being restored
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-restoring'

def test_glrestore_2(BTO):
    """
    A simple test restoring a single object to new location based on a list of files
    """
    upload_glacierized_file()

    # Make sure this file starts off glacierized
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-no-restore'

    cmd = f"glrestore -f {BTO.glacerized_file_list} --debug"
    run_glrestore(cmd)

    # Make sure this file is now being restored
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-restoring'

def test_glrestore_3(BTO):
    """
    Test the "--wait" functionality
    """
    upload_glacierized_file()

    # Make sure this file starts off glacierized
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-no-restore'

    # Make sure this takes longer than 30 seconds to finish
    t = Thread(target=threaded_restore, args=(BTO.glacerized_file_loc,))  # run the some_task function in another
    t.daemon = True
    t.start()

    sleep(30)

    # Make sure this file is now being restored
    assert glrestore.s3_utils.glacier_status_v2(BTO.glacerized_file_loc) == 'glacier-restoring'

    # Make sure the program is still running
    assert t.is_alive()

def test_glstatus_1(BTO):
    """
    A simple test of glstatus
    """
    outloc = f"{BTO.test_dir}files.csv"
    cmd = f"glrestore -f s3://sonn-current/users/mattolm/testing_house/glrestore/ --report -o {outloc}"
    print(cmd)

    run_glrestore(cmd)

    # Load the report
    cdb = pd.read_csv(outloc)
    assert len(cdb) == 3

def threaded_restore(f):
    """
    Something to thread when testing the wait
    """
    cmd = f"glrestore -f {f} --wait"
    run_glrestore(cmd)

glrestore/s3_utils.py
import boto3
import logging
import datetime
import awswrangler

import pandas as pd

from collections import defaultdict

def get_boto3_client(**kwargs):
    """
    The point of this is really to handle the "profile" option when setting up boto3
    """
    # This means you already have a client made
    if 'client' in kwargs:
        return kwargs.get('client')

    # This means I'll make you a new client
    if 'profile' in kwargs:
        profile_name = kwargs.get('profile')
        session = boto3.session.Session(profile_name=profile_name)
    else:
        session = boto3.session.Session()
    return session.client("s3")

def get_bucket_key(s3_loc):
    """
    From a full s3 location, return the bucket and key
    """
    if not s3_loc.startswith('s3://'):
        raise Exception(f"{s3_loc} is not a properly formatted key")

    bucket = s3_loc.split('/')[2]
    key = '/'.join(s3_loc.split('/')[3:])

    return bucket, key

# def get_object_storage_class(s3_loc, extra_info=False, **kwargs):
#     """
#     Return the storage class and restoring status of an s3_loc
#
#     If "extra_info", return a dictionary including creation date, last modified date, and size
#     """
#     client = get_boto3_client(**kwargs)
#     bucket, key = get_bucket_key(s3_loc)
#     re = client.head_object(Bucket=bucket, Key=key)
#
#     # Get the storage class. If STANDARD, StorageClass won't be in this
#     if ('ResponseMetadata' in re) & ('StorageClass' not in re):
#         sclass = 'STANDARD'
#     else:
#         sclass = re['StorageClass']
#
#     # Get the restore status. If not restored, 'Restore' won't be in this
#     if 'Restore' not in re:
#         rclass = False
#     else:
#         r = re['Restore']
#         active_restore = 'ongoing-request="true"' in r
#         # if not active_restore:
#         #     expiry_date = r.split('expiry-date=\"')[1]
#
#         if active_restore:
#             rclass = 'restoring'
#         else:
#             rclass = 'restored'
#
#     if extra_info:
#         retd = {'storage_class':sclass, 'restore_status':rclass}
#         re_header = re['ResponseMetadata']['HTTPHeaders']
#         for name, rname in zip(['LastModified', 'size_bytes'], ['last-modified', 'content-length']):
#             if rname in re_header:
#                 val = re_header[rname]
#
#                 if name in ['LastModified']:
#                     val = datetime.datetime.strptime(val, "%a, %d %b %Y %H:%M:%S %Z")
#
#                 retd[name] = val
#         return retd
#
#     else:
#         return sclass, rclass

def get_object_storage_class_v2(s3_locs, extra_info=False, **kwargs):
    """
    Return the storage class and restoring status of an s3_loc

    If "extra_info", return a dictionary including creation date, last modified date, and size
    """
    #assert type(s3_locs) == type([])
    f2re = awswrangler.s3.describe_objects(s3_locs)

    table = defaultdict(list)
    for f, re in f2re.items():

        # Get the storage class. If STANDARD, StorageClass won't be in this
        if ('ResponseMetadata' in re) & ('StorageClass' not in re):
            sclass = 'STANDARD'
        else:
            sclass = re['StorageClass']

        # Get the restore status. If not restored, 'Restore' won't be in this
        if 'Restore' not in re:
            rclass = False
        else:
            r = re['Restore']
            active_restore = 'ongoing-request="true"' in r
            # if not active_restore:
            #     expiry_date = r.split('expiry-date=\"')[1]

            if active_restore:
                rclass = 'restoring'
            else:
                rclass = 'restored'

        table['file'].append(f)
        table['storage_class'].append(sclass)
        table['restore_status'].append(rclass)

        re_header = re['ResponseMetadata']['HTTPHeaders']
        for name, rname in zip(['LastModified', 'size_bytes'], ['last-modified', 'content-length']):
            if rname in re_header:
                val = re_header[rname]

                if name in ['LastModified']:
                    val = datetime.datetime.strptime(val, "%a, %d %b %Y %H:%M:%S %Z")

                table[name].append(val)

    db = pd.DataFrame(table)
    db['size_bytes'] = db['size_bytes'].astype(float)
    return db

# def glacier_status(s3_loc, **kwargs):
#     """
#     Check if an object is in aws s3 glacier, and if so, return True. Else, return False.
#     """
#     sclass, rclass = get_object_storage_class(s3_loc, **kwargs)
#
#     # Object is not in glacier at all
#     if sclass not in ['GLACIER', 'DEEP_ARCHIVE']:
#         return 'no-glacier'
#
#     # Object is in glacier with no active restored
#     elif rclass is False:
#         if sclass == 'GLACIER':
#             return 'glacier-no-restore'
#         elif sclass == 'DEEP_ARCHIVE':
#             return 'deep-glacier-no-restore'
#         else:
#             print(f"WHAT!! {sclass}")
#             assert False, sclass
#
#     # Object is restored in glacier
#     elif rclass == 'restored':
#         return 'glacier-restored'
#
#     # Object is being actively restored
#     else:
#         assert rclass == 'restoring'
#         return 'glacier-restoring'

def glacier_status_v2(s3_loc, **kwargs):
    """
    Check if an object is in aws s3 glacier, and if so, return True. Else, return False.
    """
    db = get_object_storage_class_v2([s3_loc], **kwargs)
    assert len(db) == 1
    sclass = db['storage_class'].iloc[0]
    rclass = db['restore_status'].iloc[0]

    # Object is not in glacier at all
    if sclass not in ['GLACIER', 'DEEP_ARCHIVE']:
        return 'no-glacier'

    # Object is in glacier with no active restored
    elif rclass == False:
        if sclass == 'GLACIER':
            return 'glacier-no-restore'
        elif sclass == 'DEEP_ARCHIVE':
            return 'deep-glacier-no-restore'
        else:
            print(f"WHAT!! {sclass}")
            assert False, sclass

    # Object is restored in glacier
    elif rclass == 'restored':
        return 'glacier-restored'

    # Object is being actively restored
    else:
        assert rclass == 'restoring'
        return 'glacier-restoring'

# def classify_glacier_objects(files, **kwargs):
#     """
#     Classify all files
#     """
#     table = defaultdict(list)
#     for f in files:
#         table['file'].append(f)
#
#         rdic = get_object_storage_class(f, extra_info=True, **kwargs)
#         for n, v in rdic.items():
#             table[n].append(v)
#
#     db = pd.DataFrame(table)
#     db['size_bytes'] = db['size_bytes'].astype(float)
#     return db

def restore_file(f, **kwargs):
    """
    Restore "f" using the parameters in kwargs
    """
    client = get_boto3_client(**kwargs)

    obucket, okey = get_bucket_key(f)

    response = client.restore_object(
        Bucket=obucket,
        Key=okey,
        RestoreRequest={
            'Days': kwargs.get('days'),
            'GlacierJobParameters': {
             'Tier': str(kwargs.get('speed'))}})

    if kwargs.get('debug', False):
        logging.debug(response)

glrestore/glrestore.py
#!/usr/bin/env python
"""
Module Docstring
"""

from .version import __version__
__author__ = "Matt Olm"
__version__ = __version__
__license__ = "MIT"

import os
import sys
import boto3
import time
import copy
import argparse
import logging
import awswrangler
from time import sleep

import pandas as pd
import glrestore.s3_utils

def main():
    """ This is executed when run from the command line """
    args = parse_args()
    RestoreController(args).main()

class RestoreController(object):
    """
    Main controller of the restore command
    """
    def __init__(self, args):
        """
        Initialize and store args
        """
        self.args = args
        self.ori_args = copy.deepcopy(args)
        self.kwargs = vars(self.args)

    def main(self):
        """
        The main controller for restore
        """
        self.parse_arguments()

        logging.debug("Get objects to restore")
        self.file_classifications = self.get_files_to_restore_v2(self.kwargs.get('files'))

        if self.kwargs.get('report', True):
            logging.info("\n!!!!!!!!!!!\nWill NOT RESTORE anything because of --report flag; the following information is FYI only\n!!!!!!!!!!!!")

            logging.debug("Print status")
            self.print_status(sleep=False)

            logging.debug("Create report")
            self.create_report()

        else:
            logging.debug("Print status")
            self.print_status()

            logging.debug("Restoring files")
            self.restore_files()

        if self.kwargs.get('wait'):
            self.wait_for_restore()


    def parse_arguments(self):
        """
        Parse the arguments and add them this object as attributes
        """
        args = self.kwargs

        # Set up the log
        self.setup_log()

        # Set up boto3
        if 'profile' in args:
            boto3.setup_default_session(profile_name=args.get('profile'))
        else:
            session = boto3.session.Session()
            self.kwargs.client = session.client("s3")

    def get_files_to_restore_v2(self, files):
        """
        Return a list of s3 files to restore
        """
        # Get the command line argument
        base_restore = files

        # Load files if need be
        to_restore = []
        for br in base_restore:
            if not br.startswith('s3://'):
                with open(br, 'r') as r:
                    for line in r.readlines():
                        if not line.startswith('s3://'):
                            logging.error(f"CRITICAL ERROR! You passed {br} to -f, which doesn't start with s3://. I assumed this was a file of files to restore, but the line {line} also doesn't start with s3://. Will ignore {br} and {line}")
                        else:
                            to_restore.append(line.strip())
            else:
                to_restore.append(br)

        dbs = []
        for br in to_restore:
            db = glrestore.s3_utils.get_object_storage_class_v2(br)
            dbs.append(db)

        fc = pd.concat(dbs).reset_index(drop=True)
        return fc

    def print_status(self, sleep=True):
        """
        Print status and estimated costs
        """
        debug = self.kwargs.get('debug', False)

        cdb = self.file_classifications
        logging.info(f"Identified {len(cdb)} files")

        tdb = cdb[cdb['restore_status'] != False]
        logging.info(f"Of these, {len(tdb)} are being actively restored or are already restored")

        tdb = cdb[~cdb['storage_class'].isin(['GLACIER', 'DEEP_ARCHIVE'])]
        logging.info(f"Of these, {len(tdb)} are not in glacier")

        fcdb = cdb[(cdb['restore_status'] == False) & (cdb['storage_class'].isin(['GLACIER', 'DEEP_ARCHIVE']))]
        logging.info(f"Restoring the remaining {len(fcdb)} objects will cost the following:")

        self.display_restore_costs(fcdb, sleep=sleep)

        self.files_to_restore_filtered = fcdb['file'].tolist()

        if debug:
            for f in fcdb['file'].tolist():
                logging.debug(f)

    def create_report(self):
        """
        Create a report instead of actually restoring anything
        """
        outloc = self.kwargs.get('output')
        if not outloc.endswith('.csv'):
            outloc += '.csv'

        cdb = self.file_classifications
        logging.info(f"Identified {len(cdb)} files. Will create a report on them at {outloc}")
        cdb.to_csv(outloc, index=False)


    def display_restore_costs(self, fcdb, sleep=True):
        """
        Print how much this is going to cost

        NOTE- YOURE TREATING EVERYTHING AS IF IT'S BEING RESTORED FROM DEEP ARCHIVE; the "standard" is actully a bit cheaper when restoring from flexible
        """
        S3_COST_PER_GB_PER_MONTH = 0.022

        TIER2REQUEST2COST = {
            'Expedited':10,
            'Standard':0.10,
            'Bulk':0.025
        }
        TIER2REQUEST2SIZE_COST = {
            'Expedited': 0.03,
            'Standard': 0.02,
            'Bulk': 0.0025
        }

        # 0) Calculate the size and number of objects to restore
        num_obs = len(fcdb)
        size_obs = sum(fcdb['size_bytes']) / 1e9
        tier = self.kwargs.get('speed')

        # 1) Calculate the cost for the extra storage
        storage_cost = size_obs * (self.kwargs.get('days') / 30) * S3_COST_PER_GB_PER_MONTH

        # 2) Calculate the cost for the retrival costs
        t2cs = {}
        for t in ['Expedited', 'Standard', 'Bulk']:
            t2cs[t] = [(num_obs / 1000) * TIER2REQUEST2COST[t], size_obs * TIER2REQUEST2SIZE_COST[t]]

        # Display this info
        msg = "\n$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n"

        msg += f"It will cost the following to restore {num_obs} objects totalling {size_obs:.2f}GB:\n"
        for t, d in t2cs.items():
            msg += f"\t{t}: ${d[0]:.2f} + ${d[1]:.2f}\n"

        msg += f"It will also cost ${storage_cost:.2f} to restore the {size_obs:.3f}GB of data for {self.kwargs.get('days')} days"
        msg += '\n----------------------------\n'
        msg += f"Your TOTAL COST at {tier} speed will be ${storage_cost + sum(t2cs[tier]):0.2f}\n"
        msg += '\n----------------------------\n'

        if sleep:
            msg += f"You chose to restore at {tier} speed. Please quit the program now (ctrl + c) if you'd like to change that! I'll wait 5 seconds"
        msg += "\n$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$\n"

        logging.info(msg)
        if sleep:
            time.sleep(5)

    def restore_files(self):
        """
        Actually do the file restoring
        """
        files_to_restore_filtered = self.files_to_restore_filtered

        for f in files_to_restore_filtered:

            glrestore.s3_utils.restore_file(f, **self.kwargs)

        logging.info(f"Restore commands finished launching")

    def wait_for_restore(self):
        """
        Enter the loop where you wait for objects to restore before exiting the program
        """
        remaining = self.files_to_restore_filtered
        print(f"I am going to wait for {len(remaining)} files to be restored")

        start = time.time()
        while True:
            sleep(300)

            cdb = self.get_files_to_restore_v2(remaining)
            remaining = cdb[(cdb['restore_status'] == "restoring")]['file'].tolist()

            if len(remaining) == 0:
                break

            elapsed = time.time() - start

            sys.stdout.write('\r')
            # the exact output you're looking for:
            sys.stdout.write(f'Ive been waiting for {time.strftime("%Hh%Mm%Ss", time.gmtime(elapsed))}: {len(remaining)} files remain')
            sys.stdout.flush()

        elapsed = time.time() - start
        print(f'All done! The restore took {time.strftime("%Hh%Mm%Ss", time.gmtime(elapsed))}')

    def setup_log(self):
        args = self.kwargs

        # Set up the log
        root = logging.getLogger()

        if args.get('debug', False):
            root.setLevel(logging.DEBUG)
        else:
            root.setLevel(logging.INFO)

        handler = logging.StreamHandler(sys.stdout)
        if args.get('debug', False):
            root.setLevel(logging.DEBUG)
        else:
            handler.setLevel(logging.INFO)
        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        handler.setFormatter(formatter)
        root.addHandler(handler)

        # Play debug message
        logging.debug("!" * 80)
        logging.debug("Command to run was: {0}\n".format(' '.join(sys.argv)))
        logging.debug("glrestore version {0} was run \n".format(__version__))
        logging.debug("!" * 80 + '\n')


def controller(args):
    """ Main entry point of the app """
    logging.info("hello world")
    logging.info(args)

def parse_args():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '-f', '--files',
        help="File or files to be restored (or a list of files). Can include wildcards. Must start with the bucket in the format (s3://)",
        nargs='*', default=[])

    parser.add_argument(
        '-d', '--days',
        help="Number of days to restore the data to S3 for. During this time we pay for both glacier and S3 storage of the data.",
        default=7, type=int)

    parser.add_argument(
        '-s', '--speed',
        help="Speed at which to restore the data; faster is more expensive. Expedited=(1-5 min), Standard=(3-5 hr), Bulk=(12 hr)",
        default='Expedited', choices=['Expedited', 'Standard', 'Bulk'],)

    parser.add_argument(
        '--profile',
        help="AWS credential profile to use. Will use default by default")

    parser.add_argument(
        '--report',
        help='Rather than actually doing anything, just make a report of which files are matched by the -f argument and what their status is. Will make a file with this info based on the name in the -o argument',
        default=False, action="store_true")

    parser.add_argument(
        '-o', '--output',
        help='Where to store the --report information',
        default='glrestore_report.txt')

    parser.add_argument(
        '--wait',
        help='Wait for restore to finish before exiting the program. Works with --report too',
        default=False, action="store_true")

    parser.add_argument(
        '--debug',
        help='Create debugging log file',
        default=False, action= "store_true")

    parser.add_argument(
        "--version",
        action="version",
        version="%(prog)s (version {version})".format(version=__version__))

    args = parser.parse_args()
    return args

if __name__ == "__main__":
    main()

